### PR TITLE
Complete several of the utility documents for Jupyter

### PR DESCRIPTION
**Pull request by cmickeyb**
At the request of @prakashngit , combining #33 . 

### Update tests for groups

Replace invocation of the script for building a local groups and services database with one that uses the public interfaces for the services and groups databases. Move away from "host specific" configuration to a configuration based on the more general site configuration file (i.e. site.toml).

By default, testing will still occur with locally hosted services. Importing a site configuration file from another service provider (which just involves copying the site configuration file to ${PDO_HOME}/etc/sites/${SERVICE_HOST}.toml) enables use of PDO services from any location.

### Incorporate new widgets

This adds support for several utility notebooks that can be used for managing a client configuration. The notebooks themselves are included along with a collection of widgets that can be used by contract-specific notebooks. The widgets generally include ones to create lists of available resources (e.g. keys, services, groups), select amongst a list of available resources, and create or update resources.

* key-manager: manage keys available to the client
* service-manager: manage the database of services that can be used for contracts
* service-groups-manager: manage the database of service groups used for contracts

### CODE REVIEW

File: docs/notebooks/documents/getting_started.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+# %% [markdown]
+# ## How Contracts Use Jupyter
+#
+# Many of the PDO contract families provide Jupyter notebooks to simplify interaction with contract
+# objects. Commonly, the notebooks provide a factory for instantiating a notebook for each new
+# contract from one of the contract specific templates.The templates provided generally share five
+# common sections:
+#
+# * Configure the Contract Object
+# * Initialize the Interpreter and the PDO Environment
```
Comment (prakashngit) [inline docs/notebooks/documents/getting_started.py:30]: 1. not sure what is meant by configure the contract object (specifically, how is this related to initialize contract object context)
2. initialize the interpret -> you mean initialize the client?

File: docs/notebooks/documents/getting_started.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+# defines several procedures that are useful for initializing and interacting with the environment.
+#
+# In addition, the interpreter initialization configures an IPython extension that makes it easier
+# to provide code for multiple types of operations that can be performed on the contract
+# object. Specifically, the extension defines a magic keyword, `skip`, that takes a value or
+# expression that, if it evaluates to True, causes the code section to be skipped during notebook
+# execution.
+#
+# The next section in the notebook configures information about the contract object associated with
+# the notebook. In some cases, the variables will set by the notebook using the Papermill
```
Comment (prakashngit) [inline docs/notebooks/documents/getting_started.py:46]: "The next section in the notebook " and the rest of the of the documentation in this section. sounds like what "this specific notebook does" whereas I guess you are referring to notebook that is used to interact with contracts?

File: docs/notebooks/documents/getting_started.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+# [Service Manager Notebook](service_manager.ipynb)
+#
+# %%
+from pdo.contracts.services import ServiceListWidget
+from pdo.contracts.services import ServiceUploadWidget
+from pdo.contracts.services import service_labels
+
+children = map(lambda stype : ServiceListWidget(state, bindings, stype), service_labels.keys())
+service_list = ipywidgets.Tab(children=list(children), titles=list(service_labels.values()))
+ip_display.display(ipywidgets.VBox([service_list, ServiceUploadWidget(state, bindings)]))
```
Comment (prakashngit) [inline docs/notebooks/documents/getting_started.py:105]: looks like you assume that by default there  are services deployed on localhost? What if there all services are located on remote only?  I see that there is an option to upload the site.toml file, but the option is presented only after a default setting on localhost is executed.

File: docs/notebooks/documents/getting_started.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+# service will have its own service groups. PDO requires that each type of service have a
+# service group named "default" that will be used when no groups have been specified.
+#
+# The widget below contains a list of service groups for each type of service and a widget for
+# creating new services of each type. If there is no default service group, please create one.
+#
+# Additional management of service group information is provided by the [Service Groups Manager
+# Notebook](service_groups_manager.ipynb)
+# %%
+from pdo.contracts.groups import ServiceGroupListWidget
```
Comment (prakashngit) [inline docs/notebooks/documents/getting_started.py:121]: Given that there is special emphasis on the service db and serivce groups, would be better if "contract context specification" includes an option for the contract author to specify which service group to use to deploy the contract.  The Issuer notebook only specifies the service host where the contract objects will be created; what seems more meaningful is to rather specify the service groups (even if it is default) where the contract will be deployed.

File: docs/notebooks/documents/service_groups_manager.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+# ## Create a New Provisioning Service Group
+#
+# %%
+ip_display.display(ProvisioningServiceGroupCreateWidget(state, bindings))
+
+# %% [markdown]
+# ## Create a New Storage Service Group
+#
+# %%
+ip_display.display(StorageServiceGroupCreateWidget(state, bindings))
```
Comment (prakashngit) [inline docs/notebooks/documents/service_groups_manager.py:62]: (perhaps not a comment for this specific notebook:):

How do we ensure that the notebook user understands that the storage service group needs to be "compatible" with the eservice group when provisioning a contract? might be good to document this aspect somwhere.

File: docs/notebooks/documents/service_manager.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+# %% [markdown]
+# ## Verify Services
+#
+# Verification checks that the service exists and that its identity has not
+# changed since it was added to the service database. While correct operation
+# does not require service verification, it is often useful to check on the
+# status of services.
+# %%
+def verify_service(service_type, urls, output) :
```
Comment (prakashngit) [inline docs/notebooks/documents/service_manager.py:61]: "verification" term is a bit overloaded/confusing. I suppose within the esrvicedb, we verify each service against ledger, and record "ledger verified time in the db". whereas looks like in here "verify" means a "healthcheck" of the service itself that the client gets (via the /info url). how are these two verifications related?

**Review by prakashngit — COMMENTED** [review]
lots of interesting features. would have to think more about how these get used when we build a demo such as the one that are doing now.

**Comment by cmickeyb** [comment]
> lots of interesting features. would have to think more about how these get used when we build a demo such as the one that are doing now.

not intended for a demo. more intended for managing information for a pilot.

File: docs/notebooks/documents/getting_started.py (hunk 1 above)
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:30]: please suggest wording changes.

File: docs/notebooks/documents/getting_started.py (hunk 2 above)
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:46]: suggest wording changes.

File: docs/notebooks/documents/getting_started.py (hunk 3 above)
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:105]: where do you see localhost? there are no assumptions. yes, test_jupyter sets up a localhost site (that's where the services are located). but you can load any site file. i do this often for other "permanent" services.

File: docs/notebooks/documents/getting_started.py (hunk 4 above)
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:121]: that change is coming in the contracts. i have a PR ready to go that updates all of those in the exchange cf. to be clear... you will not specify a "host" but rather will specify a group.

File: docs/notebooks/documents/service_groups_manager.py (hunk 5 above)
Comment (cmickeyb) [reply docs/notebooks/documents/service_groups_manager.py:62]: but it doesn't need to be compatible. the client will load the contract to the eservice. the storage services are exclusively for your replication protocol, NOT for use in eservices. that being said... a really smart/performant policy would connect the two. but there is no requirement that they must be the same.

File: docs/notebooks/documents/service_manager.py (hunk 6 above)
Comment (cmickeyb) [reply docs/notebooks/documents/service_manager.py:61]: in this case, not really. the verify function does exactly that... it makes sure that the service exists and that the identity has not changed. this is old code.

File: docs/notebooks/documents/service_groups_manager.py (hunk 5 above)
Comment (prakashngit) [reply docs/notebooks/documents/service_groups_manager.py:62]: don't we need to assume that the storage group for sure includes sservices associated with the provisioned eservices? this would be the default assumption. We need to add more sservices to the storage group if the replication policy calls for additional redundancy.

File: docs/notebooks/documents/getting_started.py (hunk 3 above)
Comment (prakashngit) [reply docs/notebooks/documents/getting_started.py:105]: ok, when I ran the notebook, my observation was the I get the option to upload the site.toml only after executing the cell; but executing the cell meant that it automatically showed the existence of a default localhost based group. My question was what happens if such a default localhost group does not exist. I have no way to test it now, until I deploy a more elaborate set up.

File: docs/notebooks/documents/getting_started.py (hunk 4 above)
Comment (prakashngit) [reply docs/notebooks/documents/getting_started.py:121]: thanks, looking forward to that PR.

File: docs/notebooks/documents/service_manager.py (hunk 6 above)
Comment (prakashngit) [reply docs/notebooks/documents/service_manager.py:61]: ok thanks, so if I understand correctly, verify here does not verify against the ledger, just a healthcheck of the service. 
thinking aloud, for a contract user, sepcifically one who just wants to participate in an already deployed  PDO-based application (such as new token owner), what would be useful is for the user to start with the application context file, and invoke a verify API on the context file, and behind the scenes, the validity of the services specified by the whole context file gets verified (including ledger checks for attestation/registry info, service heartbeats such as what you are doing here etc.). We should not be expecting the user (I mean the new token owner) to maintain (or at least understand) any other config file other this user needs to maintain. PDO specific configs such as servicedbs, groupdbs must be constructed transparently based on the info found in context.

File: docs/notebooks/documents/getting_started.py (hunk 3 above)
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:105]: yes. we had already loaded that site.toml for the localhost test files.

File: docs/notebooks/documents/service_groups_manager.py (hunk 5 above)
Comment (cmickeyb) [reply docs/notebooks/documents/service_groups_manager.py:62]: we do not.

**Comment by g2flyer** [comment]
FYI: `make -C docker pdo_build_images` does not work anymore.  However, the issue is in `private-data-objects`. I thought initially i might be the culprit in [review of  PDO PR #477](https://github.com/hyperledger-labs/private-data-objects/pull/477#discussion_r1538098878)  but now noticed that even the old version would have failed. For the `all` target to work the sgx special build targets would seem to have to be called `build_service_sgx` and `build_client_sgx` (and not as originally build_sgx_service` or as changed due to my consistency change `sgx_build_service`).  Let me see if this quick-fix really works and i will create a corresponding PDO PR ...

**Comment by cmickeyb** [comment]
> FYI: `make -C docker pdo_build_images` does not work anymore. However, the issue is in `private-data-objects`. I thought initially i might be the culprit in [review of PDO PR #477](https://github.com/hyperledger-labs/private-data-objects/pull/477#discussion_r1538098878) but now noticed that even the old version would have failed. For the `all` target to work the sgx special build targets would seem to have to be called `build_service_sgx` (and not as originally build_sgx_service`or as changed due to my consistency change`sgx_build_service`). Let me see if this quick-fix really works and i will create a corresponding PDO PR ...

i've noticed that we are now superceding "all" as the default target. and in part its because of where we included make.loc. explicitly calling make with a target should fix this. I can add a small commit or we can just handle it separately.

**Comment by g2flyer** [comment]
> i've noticed that we are now superceding "all" as the default target. and in part its because of where we included make.loc. explicitly calling make with a target should fix this. I can add a small commit or we can just handle it separately.

[this PR](https://github.com/hyperledger-labs/private-data-objects/pull/485) should fix the issue without change here (of course whether target `all` is the right one here is a different question but `all` shouldn't fail ...)

**Review by g2flyer — CHANGES_REQUESTED** [review]
Changes look good -- definitely like the disappearance of some skip's (although still many left ;-) -- and after applying locally the pdo-build-target-fix and switching again test-machines as it obscurely fails on machines which do not support AVX -- still think my previous suggestion for a simple test would be useful .. - also builds from standalone repo.   However, running it, i run into some issues

* the inference token issuer notebook creation displays rather than renders [Newly created token issuer](../../instances/inference/user1 tokens/token-issuer_DQLVoj.ipynb), i.e., it is not clickably anymore as in earlier versions
*  pdo env initialization fails for issuer notebook due to `No such file or directory: '/project/pdo/run/opt/pdo/bin/pdo-create-service-groups.psh'`, the rest of that notebook though still worked
* the "Notebook to Invoke Token Operations" still seemed to have a number of past open comments (e.g., to show in text path to zebra image file on docker) and being too lazy to search that path, i gave up for now and wait until these prior comments are resolved

**Comment by cmickeyb** [comment]
> Changes look good -- definitely like the disappearance of some skip's (although still many left ;-) -- and after applying locally the pdo-build-target-fix and switching again test-machines as it obscurely fails on machines which do not support AVX -- still think my previous suggestion for a simple test would be useful .. - also builds from standalone repo. However, running it, i run into some issues
> 
> * the inference token issuer notebook creation displays rather than renders [Newly created token issuer](../../instances/inference/user1 tokens/token-issuer_DQLVoj.ipynb), i.e., it is not clickably anymore as in earlier versions
> * pdo env initialization fails for issuer notebook due to `No such file or directory: '/project/pdo/run/opt/pdo/bin/pdo-create-service-groups.psh'`, the rest of that notebook though still worked
> * the "Notebook to Invoke Token Operations" still seemed to have a number of past open comments (e.g., to show in text path to zebra image file on docker) and being too lazy to search that path, i gave up for now and wait until these prior comments are resolved

Presumably these issue are all in the inference notebooks. I need to go through  & update them (they did not exist when this PR was submitted). Good catch!

File: docs/notebooks/documents/getting_started.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+# object. Specifically, the extension defines a magic keyword, `skip`, that takes a value or
+# expression that, if it evaluates to True, causes the code section to be skipped during notebook
+# execution.
+#
+# The next section in the notebook configures information about the contract object associated with
+# the notebook. In some cases, the variables will set by the notebook using the Papermill
+# extensions. In some cases, the variables may be customized for the specific behavior desired.
+#
+# The following section initializes the PDO environment from the PDO configuration files. There may
+# be opportunities for customization, but so long as the PDO configuration is complete changes to
```
Comment (marcelamelara) [inline docs/notebooks/documents/getting_started.py:50]: nit
```suggestion
# be opportunities for customization, but so long as the PDO configuration is complete, changes to
```

File: docs/notebooks/documents/getting_started.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+# this section should be infrequent.
+#
+# The final common section initializes the contract context. Where the PDO initialization handles
+# configuration of the PDO modules, this section handles configuration of the specific contract
+# object and its relationship to other contract objects.
```
Comment (marcelamelara) [inline docs/notebooks/documents/getting_started.py:55]: A couple of thoughts. Are we missing a description of one of the five sections? I only see four. Also, they are out of order compared to the bulleted list above, so whichever order makes the most sense is fine, but the ordering of that list and these paragraphs should be consistent.

**Review by marcelamelara — COMMENTED** [review]
I've reviewed the introductory text so far. Will try to review the rest of this doc later.

File: docs/notebooks/documents/getting_started.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+# This document provides an overview for preparing to work with PDO contracts through the contract
+# docker container. It will walk through the necessary steps to set up a functional configuration.
+#
+# %% [markdown]
+# ## How Contracts Use Jupyter
+#
+# Many of the PDO contract families provide Jupyter notebooks to simplify interaction with contract
+# objects. Commonly, the notebooks provide a factory for instantiating a notebook for each new
+# contract from one of the contract specific templates.The templates provided generally share five
+# common sections:
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:27]: they do not. they are about the user interface that we happen to use.

File: docs/notebooks/documents/getting_started.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+#
+# * Configure the Contract Object
+# * Initialize the Interpreter and the PDO Environment
+# * Initialize the Contract Object Context
+# * Operate on the Contract
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:32]: Yes and no. The point is not to expose the methods. The point is to provide "user appropriate" functions that operate on one or in some cases several contracts to do something.

File: docs/notebooks/documents/getting_started.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+# * View Contract Metadata
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:33]: Future work.

File: docs/notebooks/documents/getting_started.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+#
+# To initialize the interpreter, the notebook loads the Juptyer helper module. This module imports
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:35]: no. it really is the interpreter.

File: docs/notebooks/documents/getting_started.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+# all of the relevant PDO and Jupyter modules to simplify access for code in the notebook. It also
+# defines several procedures that are useful for initializing and interacting with the environment.
+#
+# In addition, the interpreter initialization configures an IPython extension that makes it easier
+# to provide code for multiple types of operations that can be performed on the contract
+# object. Specifically, the extension defines a magic keyword, `skip`, that takes a value or
+# expression that, if it evaluates to True, causes the code section to be skipped during notebook
+# execution.
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:43]: to the first... um... yeah. jupyter notebooks are interpreter instances. they are interpreted python and you can do pretty much anything you want in the notebook cells. the security model for jupyter is basically local access only with password or token access. its more or less like running any other interactive shell.

to the second... yes. add an issue?

File: docs/notebooks/documents/getting_started.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:46]: this section has been rewritten.

File: docs/notebooks/documents/getting_started.py (hunk 8 above)
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:55]: yes. fixed.

File: docs/notebooks/documents/getting_started.py (hunk 1 above)
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py]: Section rewritten. However, this really is about the Jupyter interpreter and the PDO Environment.

File: docs/notebooks/documents/getting_started.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+# common sections:
+#
+# * Configure the Contract Object
+# * Initialize the Jupyter Interpreter and the PDO Environment
+# * Initialize the Contract Object Context
+# * Operate on the Contract
+# * View Contract Metadata
+#
+# *Configure the Contract Object*: Each notebook either loads an existing contract object from a
+# file (see the PDO documentation on contract collections) or creates a new contract object based on
```
Comment (prakashngit) [inline docs/notebooks/documents/getting_started.py:38]: I think "Configure the Contract Object" is perhaps misleading: We are "configuring the notebook" (using the factory notebook as a base) as required to interact with the contract

File: docs/notebooks/documents/getting_started.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+#
+# *Initialize the Contract Object Context*: The next section in the notebook creates a PDO context
+# for the contract object (see the PDO documentation for more information on contexts).  The context
+# includes all information necessary to create the contract object (and its dependencies) based on
+# the initial set of parameters provided by the factory notebook. Beyond the initial parameters, the
+# context allows for fine-grained customization of the contract object, though the defaults are
+# usually sufficient.
+#
+# *Operate on the Contract*: The contract object may be used once it has been created and
+# initialized. In general, cells in this section of the notebook are turned off by default; that is,
```
Comment (prakashngit) [inline docs/notebooks/documents/getting_started.py:61]: "Initialize the Contract Object Context" -> Perhaps change this to "Initialize the application context" by specifying the contracts invovled in the application, and also the inter-dependency among the contracts used by the applicaiton.

File: docs/notebooks/documents/getting_started.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+# `%%skip True` is added at the top of the cell. To perform an operation, change the top line to
+# `%%skip False` and evaluate the cell.
```
Comment (prakashngit) [inline docs/notebooks/documents/getting_started.py:63]: Same comment as above "Operate on the contract" -> Perhaps change this to "operate on the app"

Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:63]: > Same comment as above "Operate on the contract" -> Perhaps change this to "operate on the app"

btw... i really hope you've spent this much time on the other new notebooks out there.

Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:63]: while i don't disagree... the metaphor for a notebook (app? contract?) runs through much of our documentation. can i suggest that rather than microanalyze it in one document, that we step back and address the bigger question of terminology (and consistency with it) in an issue? i'm tempted to completely remove this section & leave this document focused on operational configuration.

File: docs/notebooks/documents/getting_started.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (marcelamelara) [reply docs/notebooks/documents/getting_started.py:34]: ah I see, so this is really a level of abstraction above individual contract method invocations? The other concern I have with "operate on a contract" is that, to me, it sounds a lot like configuration, so I would suggest making the distinction between the config step and the operate step a bit more explicit.

File: docs/notebooks/documents/getting_started.py (hunk 14 above)
Comment (marcelamelara) [inline docs/notebooks/documents/getting_started.py:38]: Nit: can we add a link to the referenced documentation? It'll be much easier for the reader to find the right information if we provide the link directly.

File: docs/notebooks/documents/getting_started.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+# are useful for initializing and interacting with the PDO environment.
+#
+# In addition, the Jupyter interpreter initialization configures an IPython extension that makes it
+# easier to provide code for multiple types of operations that can be performed on the contract
+# object. Specifically, the extension defines a magic keyword, `skip`, that takes a value or
+# expression that, if it evaluates to True, causes the code section to be skipped during notebook
+# execution.
+#
+# *Initialize the Contract Object Context*: The next section in the notebook creates a PDO context
+# for the contract object (see the PDO documentation for more information on contexts).  The context
```
Comment (marcelamelara) [inline docs/notebooks/documents/getting_started.py:54]: Same as above, if the documentation already exists, we should add a link here.

File: docs/notebooks/documents/getting_started.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+
+key_list = PrivateKeyListWidget(state, bindings)
+key_gen = GenerateKeyWidget(state, bindings)
+ip_display.display(ipywidgets.VBox([key_list, key_gen]))
+
+# %% [markdown]
+# ## Import Site Information
+#
+# PDO clients require information about the services that will be used for contracts. Typically, a
+# service provider will create a site description file (often called `site.toml`) that can be used
```
Comment (marcelamelara) [inline docs/notebooks/documents/getting_started.py:114]: What are the assumptions about where PDO clients can obtain site.toml files to import?

File: docs/notebooks/documents/getting_started.py [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+# %% [markdown]
+# ## Create Service Groups
+#
+# Service groups are a shortcut for configuration of collections of enclave, storage, and
+# provisioning services used to create and provision a PDO contract object. Each type of
+# service will have its own service groups. PDO requires that each type of service have a
+# service group named "default" that will be used when no groups have been specified.
+#
+# The widget below contains a list of service groups for each type of service and a widget for
+# creating new services of each type. If there is no default service group, please create one.
```
Comment (marcelamelara) [inline docs/notebooks/documents/getting_started.py:139]: Is there a way to link to `service_groups_manager.py` from here?

**Review by marcelamelara — COMMENTED** [review]
Thanks for the edits @cmickeyb . I mostly have comments about making existing documentation easier to find at this point.

File: docs/notebooks/documents/getting_started.py (hunk 19 above)
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:139]: Yes. thought i did. will fix.

File: docs/notebooks/documents/getting_started.py (hunk 14 above)
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:38]: not  really. the only way to reference the PDO documentation would be through the git repository.

File: docs/notebooks/documents/getting_started.py (hunk 17 above)
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:54]: agreed... and if you have a suggestion for how to do that I would happily incorporate. the closest that we have is to put in a ref to the pdo git hub repo. not my favorite solution.

File: docs/notebooks/documents/getting_started.py (hunk 18 above)
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.py:114]: we have none per se. for the moment, my preference is to keep that "out of band".

**Review by prakashngit — APPROVED** [review]
Thanks, tests work as long as pdo images have been generated. if you are using the  pdo submodule referenced to by this PR, please note that pdo images need to be generated directly from within the pdo folder, rather than using the `make build_pdo_images`. This fix for this pdo-related (and not specific to this PR) appears in future pdo commits. This will become a non-issue once we update the pdo submodule for this contracts in later PR.

File: docs/notebooks/documents/getting_started.py [hunk 20]
```diff
@@ -13,64 +13,15 @@
...
 
 # %% [markdown]
-# # Getting Started
+# # Managing the PDO Environment
 #
-# This document provides an overview for preparing to work with PDO contracts through the contract
-# docker container. It will walk through the necessary steps to set up a functional
-# configuration. It does assume that the reader is familiar with
+# This notebook provides an aggregate interface for preparing to work with PDO contracts through the
+# necessary steps to set up a functional configuration. It assumes that the reader is familiar with
```
Comment (marcelamelara) [inline docs/notebooks/documents/getting_started.py:19]: The first sentence is a little convoluted. 
```suggestion
# This notebook provides an aggregate interface for preparing to work with PDO contracts, and describes the
# necessary steps to set up a functional configuration. It assumes that the reader is familiar with
```

File: docs/notebooks/documents/getting_started.py (hunk 18 above)
Comment (marcelamelara) [reply docs/notebooks/documents/getting_started.py:72]: Alright, that's fine. I always like to make assumptions like these explicit (if they aren't already somewhere) to avoid any confusion.

**Review by marcelamelara — APPROVED** [review]
Thanks! Just one small edit to the text at the top.

**Review by g2flyer — APPROVED** [review]
Approve as `make -C test` works and inference contract issues i pointed out in https://github.com/hyperledger-labs/pdo-contracts/pull/32#pullrequestreview-2007629098 are deferred to PR #37
